Wrap weighted histograms

MikeSpreitzer opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?

<!--
Add one of the following kinds:
/kind bug
/kind cleanup
/kind documentation
/kind feature

Optionally add one or more of the following kinds if applicable:
/kind api-change
/kind deprecation
/kind failing-test
/kind flake
/kind regression
-->
/kind cleanup

#### What this PR does / why we need it:
This PR adds kube wrappers around the new WeightedHistograms and TimingHistograms (introduced in #109277), along the lines of the existing wrappers.

#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
This partially addresses #108272

#### Special notes for your reviewer:

This currently builds on #109277 , and after that merges I will rebase this onto master.
This is the middle of three steps.
The last step will be replacing the sample-and-watermark histograms with uses of the timing histograms.

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```

MikeSpreitzer committed:
Add wrapper for TimingHistogram

Do not bother wrapping WeightedHistogram because it is not used
in k/k.

MikeSpreitzer commented:
/sig instrumentation
/sig architecture

MikeSpreitzer commented:
/cc @wojtek-t 
/cc @dgrisonnet

MikeSpreitzer commented:
/retest

MikeSpreitzer commented:
/retest

wojtek-t commented on staging/src/k8s.io/component-base/metrics/weighted_histogram.go:
```diff
@@ -0,0 +1,213 @@
+
+// WeightedHistogram is our internal representation for our wrapping struct around prometheus
+// histograms. It implements both kubeCollector and ObserverMetric
+type WeightedHistogram struct {
```
Do we have a usecase for using WeightedHistogram in k8s codebase?
My understanding is that for now we don't have any - and if that's true, let's not expose that for now (and treat it purely as an implementation detail for TimingHistogram from k8s perspective).

[less code = less maintenance]

MikeSpreitzer replied:
done

MikeSpreitzer commented:
The force-push to https://github.com/kubernetes/kubernetes/commit/7379d1c0112dc7c1601378c032daa089bf28edc1 is a rebase onto master now that that includes the prometheus extensions.

MikeSpreitzer commented:
/retest

MikeSpreitzer commented:
/retest

MikeSpreitzer commented:
/test pull-kubernetes-integration

wojtek-t commented on staging/src/k8s.io/component-base/metrics/timing_histogram_test.go:
```diff
@@ -0,0 +1,371 @@
+			assert.Nil(t, err, "Gather failed %v", err)
+
+			for _, metric := range ms {
+				expectedHelp := helpPrefix + test.expectedHelp
```
Instead of playing with setting expectedHelp here, please set the expectedHelp field in all structs appropriate to have the EXPERIMENAL: prefix.

[Same for vector test.]

wojtek-t reviewed (commented):
/assign

Just one minor NIT from me - other than that it LGTM.

/approve

I will let @dgrisonnet - to make a final stamp though (or at least confirm he's fine with it).
/assign @dgrisonnet

logicalhan commented:
/triage accepted
/assign @dgrisonnet @logicalhan

MikeSpreitzer commented:
The force-push to 32a9dbfbde8 addresses the review comments above.

MikeSpreitzer commented:
The force-push to 18d9a38c0b9 adds a little overlooked cleanup.

MikeSpreitzer commented:
af88f5762b0 adds a little more overlooked cleanup.

MikeSpreitzer commented:
74e1f768e5b is still more cleanup

MikeSpreitzer commented:
The force-push to 68d92494907 adds Reset to the methods of GaugeMetricVec, because why were we keeping it out anyway?

MikeSpreitzer committed:
A bit more tidying up

MikeSpreitzer commented:
/retest

MikeSpreitzer commented:
A brief conversation with @logicalhan turned up no enthusiasm for changing all the wrappers, so I will follow the existing pattern.  The force-push to b2c0c22e8f4 removes the remodeling.

wojtek-t commented on staging/src/k8s.io/component-base/metrics/wrappers.go:
```diff
@@ -65,6 +68,70 @@ type GaugeMetric interface {
+// GaugeMetricVec is a collection of Gauges that differ only in label values.
+// This is really just one Metric.
+// It might be better called GaugeVecMetric, but that pattern of name is already
+// taken by the other pattern --- which is treacherous.  The treachery is that
```
Taken by what?
I see "CounterVecMetric" which seems better than what you have here.

wojtek-t replied:
friendly ping

wojtek-t commented on staging/src/k8s.io/component-base/metrics/wrappers.go:
```diff
@@ -65,6 +68,70 @@ type GaugeMetric interface {
+	Add(delta float64, labelValues ...string)
+	SetToCurrentTime(labelValues ...string)
+
+	SetForLabels(value float64, labels map[string]string)
```
Do we really need those methods? I don't see them implemented by e.g. HistogramVec, and so I don't see a reason why we really would need them here. I really prefer consistency...

MikeSpreitzer commented:
@wojtek-t , thanks for your latest review.  All your comments are about the same issue, so I will address it here coherently.  Both the HistogramVec and CounterVec wrappers have the problem, since they both follow the same pattern.  One critical part of the problem is in the code that extracts a Vec member.  It starts like this:

```
func (v *CounterVec) WithLabelValues(lvs ...string) CounterMetric {
	if !v.IsCreated() {
		return noop // return no-op counter
	}
```

The Vec is not "created" until it is registered in at least one registry.  The `WithLabelValues` method is preceded by a comment that is not wrong, but did not make the important consequence blindingly obvious to me.  That consequence is that if the client holds on to the result of this call then the client will have a permanent noop.

Normally that is not a problem, because the common pattern for using such a Vec is code like the following.

```
    myVec.WithLabelValues("v1", "v2").Inc()
```

That code will do a no-op whenever it runs before `myVec` is registered, and increment an actual counter after the registration.

Also, that code has the overhead of extracting the relevant member from the Vec, and constructing an object that escapes to the heap, for every increment.

In APF we have SampleAndWatermark _client_ code that takes the opposite approach.  The SampleAndWatermark code currently uses the word "Generator" instead of "Vec", but it is essentially the same thing (and you can see that this PR switches to using "Vec" at that layer).  The SampleAndWatermark _client_ extracts the relevant members at package initialization time, which is before the relevant registration.  It works for SampleAndWatermark histograms because their _implementation_ is layered on ordinary wrapped histograms and uses members according to the pattern above. 

This PR proposes to replace SampleAndWatermark histogram "generators" with TimingRatioHistogram "vectors" and follow the usual vector pattern.  But simply making the TimingRatioHistogram implementation follow exactly the usual vector pattern breaks the client layer (the maxinflight.go and priority-and-fairness.go filters), which now latches onto permanent noops at package initialization time. So I had a few thoughts about improvements.

1. The usual pattern for using a vector member has an unnecessary heap object creation and collection.  We get more efficient code if we combine the member extraction with the member use into a single method that avoids the unnecessary heap thrashing.
2. The usual pattern for using a vector member has unnecessary computation in the cases where the call site is using constant label values (either totally constant, or constant per some other object at hand).  Better to extract the vector member up front, to avoid unnecessarily repeating the computation.
3. The existing method for extracting a vector member is treacherous --- it sometimes returns a noop without letting the client know about it.  Better to let the client know.  So I extended WithLabelValues and WithLabel with a way for the client to know whether it got shortchanged due to registration not happening yet.
4. In some cases a call site that wants to use a single Vec member and knows which one before registration is willing to pay the cost of the intermediate heap object.  It is neater to capture an object that implements the member interface at initialization time.  Even neater if this member extraction will return the more efficient post-registration member if called post registration (which it discerns by looking at the extended return from WithLabelValues).  That is what the WithLabelValuesSafe method is for.

MikeSpreitzer committed:
Make more ordinary and add benchmarks of wrapped timing histograms

MikeSpreitzer commented:
I did a round of revisions, with an eye toward making the usual pattern available.

I removed the combined methods (e.g., `Set(float64, ...string)`) because it is not more efficient at this layer: WithLabelValues does not create an additional heap object and the rest of the member extraction costs are paid regardless.

I changed names so that the name `WithLabelValues` has the usual signature and semantics.  The additional method `WithLabelValuesChecked` exposes the extended semantics.

wojtek-t commented on staging/src/k8s.io/component-base/metrics/timing_histogram.go:
```diff
@@ -0,0 +1,253 @@
+// If none of the above apply, this method will return
+// the appropriate vector member and a nil error.
+func (v *timingHistogramVec) WithLabelValuesChecked(lvs ...string) (GaugeMetric, error) {
+	if v.IsHidden() {
```
OK - thanks for the explanation in:
https://github.com/kubernetes/kubernetes/pull/109729#issuecomment-1121353342

That makes sense to me and I'm fine with that.
But I want @dgrisonnet or @logicalhan to chime in here too.

dgrisonnet replied:
The approach taken to noop the metrics until they are initialized is correct. The deferred initialization of the metrics is normal behavior since the Kubernetes version is held by the Regsitry and because of that, it is impossible to determine if a metric is deprecated until it is registered.

As per handling whether the metrics is Hidden or not, that shouldn't be necessary since this is already taken care of when creating the underlying metric.
- some preprocessing will determine if the metric should be hidden: https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/component-base/metrics/metric.go#L90-L100
- after this preprocessing, if the metrics is marked as hidden, it will not be created: https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/component-base/metrics/metric.go#L146-L149

wojtek-t commented on staging/src/k8s.io/component-base/metrics/wrappers.go:
```diff
@@ -65,6 +68,78 @@ type GaugeMetric interface {
+	Reset()
+}
+
+// PreContextGaugeMetricVec is something that can construct a GaugeMetricVec
```
As suggested above - can we get rid of those interfaces? I think they add additional complexity for lack of clear value.

MikeSpreitzer committed:
Finish making Gauge and TimingHistogram implement the same interface

MikeSpreitzer commented on staging/src/k8s.io/component-base/metrics/timing_histogram.go:
```diff
@@ -0,0 +1,253 @@
+// NewTimingHistogramVec returns an object which satisfies kubeCollector and
+// wraps the promext.timingHistogramVec object.  Note well the way that
+// behavior depends on registration and whether this is hidden.
+func NewTimingHistogramVec(opts *TimingHistogramOpts, labels []string) PreContextAndRegisterableGaugeMetricVec {
```
With the alignment of WithLabelValues and of With, Gauge and TimingHistogram now implement the same interfaces, as shown in the latest revision.

MikeSpreitzer replied:
The fact that these two things implement the same interface is an important insight from @beorn7 , and I think the code should clearly say that.

wojtek-t commented on staging/src/k8s.io/component-base/metrics/counter.go:
```diff
@@ -106,9 +107,14 @@ type CounterVec struct {
 
+var _ kubeCollector = &CounterVec{}
+
+// var _ CounterVecMetric = &CounterVec{}
```
Uncomment or remove

wojtek-t commented on staging/src/k8s.io/component-base/metrics/wrappers.go:
```diff
@@ -65,6 +68,94 @@ type GaugeMetric interface {
+type PreContextGaugeMetric interface {
+	// WithContext creates a GaugeMetric that uses the given Context
+	WithContext(ctx context.Context) GaugeMetric
+}
```
This type isn't used anywhere else that the interface below. I'm somewhat against introducing interfaces for the sake of introducing them. Why can't we inline it below?

wojtek-t commented on staging/src/k8s.io/component-base/metrics/wrappers.go:
```diff
@@ -65,6 +68,94 @@ type GaugeMetric interface {
+}
+
+// PreContextAndRegisterableGaugeMetric extends GaugeMetric with registration, collection, and context binding
+type PreContextAndRegisterableCollectorGaugeMetric interface {
```
> The fact that these two things implement the same interface is an important insight from [@beorn7](https://github.com/beorn7) , and I think the code should clearly say that.

I don't think it convinces me to actually buy those hard to follow interfaces.
What really matters in my opinion is the fact that TimingHistogram has exactly the same interface as GaugeMetric.
The fact of registerable, kubeCollector etc. are implementation details.

So I agree that we need conditions like:
```
var _ GaugeMetric = &TimingHistogram{}
var _ GaugetVecMetric = &TimingHistogramVec{}
```

and the need to GaugeVecMetric interface below, but I don't buy the need for any other interface here.
They introduce more congnitive burden than gain.

MikeSpreitzer commented:
@wojtek-t : revised again
@dgrisonnet : can you please take a look?

MikeSpreitzer commented on staging/src/k8s.io/component-base/metrics/gauge.go:
```diff
@@ -130,40 +140,67 @@ func (v *GaugeVec) initializeDeprecatedMetric() {
 
-// Default Prometheus behavior actually results in the creation of a new metric
-// if a metric with the unique label values is not found in the underlying stored metricMap.
+func (v *GaugeVec) WithLabelValuesChecked(lvs ...string) (GaugeMetric, error) {
```
It is not just an implementation detail, some clients want to know about the error.  #109742 has an example.

dgrisonnet replied:
sounds good, I was wondering why you wanted to add new methods, but it makes sense to have them

MikeSpreitzer commented on staging/src/k8s.io/component-base/metrics/gauge.go:
```diff
@@ -130,40 +140,67 @@ func (v *GaugeVec) initializeDeprecatedMetric() {
+	panic(err)
+}
+
+func (v *GaugeVec) WithChecked(labels map[string]string) (GaugeMetric, error) {
```
It is not just an implementation detail, some clients want to know about the error. https://github.com/kubernetes/kubernetes/pull/109742 has an example.

MikeSpreitzer commented on staging/src/k8s.io/component-base/metrics/timing_histogram.go:
```diff
@@ -0,0 +1,261 @@
+// TimingHistogram is our internal representation for our wrapping struct around timing
+// histograms. It implements both kubeCollector and GaugeMetric
+type TimingHistogram struct {
+	GaugeMetric
```
The `promext.TimingHistogram` interface has additional methods that would be exposed, by virtue of extending `Collector`.

MikeSpreitzer replied:
I just pushed a revision that splits the difference, introducing an interface that defines what aspects of the underlying TimingHistogram to promote.

wojtek-t commented:
I think I'm roughly fine with it, but @dgrisonnet seems to have objections in some places so those needs to be resolved first (he is more familiar with this codebase than I am).

MikeSpreitzer committed:
More revisions after review

MikeSpreitzer commented on staging/src/k8s.io/component-base/metrics/timing_histogram.go:
```diff
@@ -0,0 +1,261 @@
+}
+
+// setPrometheusHistogram sets the underlying KubeGauge object, i.e. the thing that does the measurement.
+func (h *TimingHistogram) setPrometheusHistogram(histogram promext.TimingHistogram) {
```
Being an interface, taking a pointer is not something to do.

dgrisonnet replied:
indeed, I mistook it for a struct

MikeSpreitzer commented:
@dgrisonnet , I pushed a revision that responds to your review.

The revision made github stop showing the discussion around the handling of hidden vector members.  I am not sure I exactly understand your point, but thanks for catching that the hidden bit is not defined before creation.  I deliberately want to treat created and hidden differently from not created, because I think they arise in different situations and should be handled differently (again, refer to the higher level code in #109742).  The interface comments are in terms of registration rather than creation, because I think registration is the more prominent thing to clients.

MikeSpreitzer commented:
@dgrisonnet : thanks for the reviews!

MikeSpreitzer commented:
/retest

wojtek-t commented:
Just for posterity - the most recent changes LGTM.
